[BUG] ClassCastException in JsonSchemaElementUtils on fields with polymorphic subtypes

**Describe the bug**

`JsonSchemaElementUtils.getActualType` casts the single type argument of a `Collection<E>`-shaped field to `Class` unconditionally. 
The cast was introduced in c6f4277916 (Oct 2024) and stayed dormant until #5060 added `polymorphicSchemaFrom`. Before #5060 the walker treated polymorphic types as opaque, so subtype fields were never visited. After #5060 the walker descends into every concrete subtype and walks its declared fields, which surfaces the latent cast.


**Log and Stack trace**

```
Caused by: java.lang.ClassCastException: class sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl
cannot be cast to class java.lang.Class
at dev.langchain4j.internal.JsonSchemaElementUtils.getActualType(JsonSchemaElementUtils.java:327)
at dev.langchain4j.internal.JsonSchemaElementUtils.jsonSchemaElementFrom(JsonSchemaElementUtils.java:89)   
at dev.langchain4j.internal.JsonSchemaElementUtils.jsonObjectOrReferenceSchemaFrom(JsonSchemaElementUtils.java:269)
at dev.langchain4j.internal.JsonSchemaElementUtils.polymorphicSchemaFrom(JsonSchemaElementUtils.java:122)  
at dev.langchain4j.internal.JsonSchemaElementUtils.jsonSchemaElementFrom(JsonSchemaElementUtils.java:95)
at dev.langchain4j.internal.JsonSchemaElementUtils.jsonObjectOrReferenceSchemaFrom(JsonSchemaElementUtils.java:269)
at dev.langchain4j.internal.JsonSchemaElementUtils.jsonSchemaElementFrom(JsonSchemaElementUtils.java:98)
at dev.langchain4j.agent.tool.ToolSpecifications.toolSpecificationFrom(ToolSpecifications.java:112)
at dev.langchain4j.service.tool.ToolService.findTools(ToolService.java:170)
at dev.langchain4j.service.AiServices.tools(AiServices.java:477)
```

**To Reproduce**

Any `@Tool` parameter where a polymorphic field `@JsonSubTypes` has a concrete subtype declaring a `Collection<E>` field whose element type is itself parameterized. Minimal example:

```java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "kind")
@JsonSubTypes(@JsonSubTypes.Type(value = MySub.class, name = "sub"))
abstract class Base { }

class MySub extends Base {
    List<Box<String>> items;   // element type is ParameterizedType, not Class
}

class Box<T> { T value; }

class MyTools {
    @Tool void doStuff(Base input) { }
}

AiServices.builder(MyService.class).chatModel(model).tools(new MyTools()).build();
//  ClassCastException at construction time
```

**Expected behavior**

Schema generation should not crash on these field shapes. The element type can be reduced to a usable raw Class (Box, List, ...) so the walker can produce a sensible schema; it does not need to model the inner type argument.

**Please complete the following information:**
- LangChain4j version: 1.14.x and the current 1.15.0-SNAPSHOT main
- LLM(s) used: any (fails before the model is ever called)
- Java version: 17+
- Spring Boot version (if applicable): n/a

**Additional context**

PR with fix and regression tests will follow.

--- a/langchain4j-core/src/main/java/dev/langchain4j/internal/JsonSchemaElementUtils.java
+++ b/langchain4j-core/src/main/java/dev/langchain4j/internal/JsonSchemaElementUtils.java
@@ -117,15 +117,15 @@ public class JsonSchemaElementUtils {
         }
 
         String reference = generateUUIDFrom(baseType.getName());
-        VisitedClassMetadata metadata =
-                new VisitedClassMetadata(JsonReferenceSchema.builder().reference(reference).build(), reference, false);
+        VisitedClassMetadata metadata = new VisitedClassMetadata(
+                JsonReferenceSchema.builder().reference(reference).build(), reference, false);
         visited.put(baseType, metadata);
 
         String discriminatorProperty = discriminatorPropertyName(baseType);
         List<JsonSchemaElement> options = new ArrayList<>();
         for (Class<?> subtype : findConcreteSubtypes(baseType)) {
-            JsonSchemaElement subtypeSchema = jsonObjectOrReferenceSchemaFrom(
-                    subtype, null, areSubFieldsRequiredByDefault, visited, false);
+            JsonSchemaElement subtypeSchema =
+                    jsonObjectOrReferenceSchemaFrom(subtype, null, areSubFieldsRequiredByDefault, visited, false);
             JsonSchemaElement withDiscriminator =
                     addDiscriminator(subtypeSchema, baseType, subtype, discriminatorProperty);
             options.add(withDiscriminator);
@@ -135,9 +135,11 @@ public class JsonSchemaElementUtils {
                 subtypeMetadata.jsonSchemaElement = withDiscriminator;
             }
         }
-        String desc = description != null ? description : Optional.ofNullable(descriptionFrom(baseType))
-                .orElse(baseType.getSimpleName());
-        JsonAnyOfSchema anyOf = JsonAnyOfSchema.builder().description(desc).anyOf(options).build();
+        String desc = description != null
+                ? description
+                : Optional.ofNullable(descriptionFrom(baseType)).orElse(baseType.getSimpleName());
+        JsonAnyOfSchema anyOf =
+                JsonAnyOfSchema.builder().description(desc).anyOf(options).build();
         metadata.jsonSchemaElement = anyOf;
         return anyOf;
     }
@@ -162,8 +164,7 @@ public class JsonSchemaElementUtils {
             JsonTypeInfo info = baseType.getAnnotation(JsonTypeInfo.class);
             // The discriminator field is allowed to coexist with a same-named bean field only when
             // @JsonTypeInfo(visible=true) or @JsonTypeInfo(include=As.EXISTING_PROPERTY).
-            boolean allowed = info != null
-                    && (info.visible() || info.include() == JsonTypeInfo.As.EXISTING_PROPERTY);
+            boolean allowed = info != null && (info.visible() || info.include() == JsonTypeInfo.As.EXISTING_PROPERTY);
             if (!allowed) {
                 throw new IllegalArgumentException(String.format(
                         "Polymorphic subtype %s declares a field named '%s', which collides with the discriminator "
@@ -171,16 +172,14 @@ public class JsonSchemaElementUtils {
                                 + "name with @JsonTypeInfo(property = \"...\") on %s, set @JsonTypeInfo(visible = true), "
                                 + "or use @JsonTypeInfo(include = As.EXISTING_PROPERTY) if the field is intentionally "
                                 + "part of the subtype.",
-                        subtype.getName(),
-                        discriminatorProperty,
-                        baseType.getName(),
-                        baseType.getName()));
+                        subtype.getName(), discriminatorProperty, baseType.getName(), baseType.getName()));
             }
         }
 
         Map<String, JsonSchemaElement> properties = new LinkedHashMap<>();
         properties.put(
-                discriminatorProperty, JsonEnumSchema.builder().enumValues(discriminatorValue).build());
+                discriminatorProperty,
+                JsonEnumSchema.builder().enumValues(discriminatorValue).build());
         obj.properties().forEach(properties::putIfAbsent);
 
         List<String> required = new ArrayList<>();

--- a/langchain4j-core/src/main/java/dev/langchain4j/internal/JsonSchemaElementUtils.java
+++ b/langchain4j-core/src/main/java/dev/langchain4j/internal/JsonSchemaElementUtils.java
@@ -26,8 +26,11 @@ import dev.langchain4j.model.chat.request.json.JsonSchemaElement;
 import dev.langchain4j.model.chat.request.json.JsonStringSchema;
 import dev.langchain4j.model.output.structured.Description;
 import java.lang.reflect.Field;
+import java.lang.reflect.GenericArrayType;
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.lang.reflect.WildcardType;
 import java.math.BigDecimal;
 import java.math.BigInteger;
 import java.util.*;
@@ -84,9 +87,10 @@ public class JsonSchemaElementUtils {
         }
 
         if (Collection.class.isAssignableFrom(clazz)) {
+            Type elementType = collectionElementType(type);
             return JsonArraySchema.builder()
                     .items(jsonSchemaElementFrom(
-                            getActualType(type), null, null, areSubFieldsRequiredByDefault, visited))
+                            rawClassOf(elementType), elementType, null, areSubFieldsRequiredByDefault, visited))
                     .description(fieldDescription)
                     .build();
         }
@@ -320,14 +324,53 @@ public class JsonSchemaElementUtils {
         return String.join(" ", description.value());
     }
 
-    private static Class<?> getActualType(Type type) {
+    /**
+     * Returns the single type argument of a {@code Collection<E>}-shaped type, preserving its full
+     * generic shape so callers can recurse into nested generics like {@code List<List<X>>} or {@code List<Foo<Bar>>}.
+     */
+    private static Type collectionElementType(Type type) {
         if (type instanceof final ParameterizedType parameterizedType) {
             Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
             if (actualTypeArguments.length == 1) {
-                return (Class<?>) actualTypeArguments[0];
+                return actualTypeArguments[0];
             }
         }
         return null;
+    }
+
+    /**
+     * Reduces an arbitrary {@link Type} to a usable {@link Class}.
+     * Handles the common cases that arise when walking generic field types, e.g. a
+     * {@code Collection<X<Y>>} field has an actual type argument of {@code X<Y>}, which is a
+     * {@link ParameterizedType}, not a {@link Class}.
+     */
+    private static Class<?> rawClassOf(Type type) {
+        if (type == null) {
+            return Object.class;
+        }
+        if (type instanceof Class<?> clazz) {
+            return clazz;
+        }
+        if (type instanceof ParameterizedType parameterizedType
+                && parameterizedType.getRawType() instanceof Class<?> raw) {
+            return raw;
+        }
+        if (type instanceof WildcardType wildcardType) {
+            Type[] upperBounds = wildcardType.getUpperBounds();
+            if (upperBounds.length > 0) {
+                return rawClassOf(upperBounds[0]);
+            }
+        }
+        if (type instanceof TypeVariable<?> typeVariable) {
+            Type[] bounds = typeVariable.getBounds();
+            if (bounds.length > 0) {
+                return rawClassOf(bounds[0]);
+            }
+        }
+        if (type instanceof GenericArrayType) {
+            return Object[].class;
+        }
+        return Object.class;
     }
 
     static boolean isCustomClass(Class<?> clazz) {

--- a/langchain4j-core/src/main/java/dev/langchain4j/internal/JsonSchemaElementUtils.java
+++ b/langchain4j-core/src/main/java/dev/langchain4j/internal/JsonSchemaElementUtils.java
@@ -6,7 +6,6 @@ import static dev.langchain4j.internal.PolymorphicTypes.findConcreteSubtypes;
 import static dev.langchain4j.internal.PolymorphicTypes.isPolymorphic;
 import static dev.langchain4j.internal.PolymorphicTypes.verifyJsonTypeInfoIsSupported;
 import static dev.langchain4j.internal.Utils.generateUUIDFrom;
-import static java.lang.reflect.Array.*;
 import static java.lang.reflect.Modifier.isStatic;
 import static java.util.Arrays.stream;
 
@@ -368,8 +367,7 @@ public class JsonSchemaElementUtils {
             }
         }
         if (type instanceof GenericArrayType genericArrayType) {
-            Class<?> componentClass = rawClassOf(genericArrayType.getGenericComponentType());
-            return newInstance(componentClass, 0).getClass();
+            return rawClassOf(genericArrayType.getGenericComponentType()).arrayType();
         }
         return Object.class;
     }

--- a/langchain4j-core/src/main/java/dev/langchain4j/internal/JsonSchemaElementUtils.java
+++ b/langchain4j-core/src/main/java/dev/langchain4j/internal/JsonSchemaElementUtils.java
@@ -6,6 +6,7 @@ import static dev.langchain4j.internal.PolymorphicTypes.findConcreteSubtypes;
 import static dev.langchain4j.internal.PolymorphicTypes.isPolymorphic;
 import static dev.langchain4j.internal.PolymorphicTypes.verifyJsonTypeInfoIsSupported;
 import static dev.langchain4j.internal.Utils.generateUUIDFrom;
+import static java.lang.reflect.Array.*;
 import static java.lang.reflect.Modifier.isStatic;
 import static java.util.Arrays.stream;
 
@@ -367,8 +368,9 @@ public class JsonSchemaElementUtils {
                 return rawClassOf(bounds[0]);
             }
         }
-        if (type instanceof GenericArrayType) {
-            return Object[].class;
+        if (type instanceof GenericArrayType genericArrayType) {
+            Class<?> componentClass = rawClassOf(genericArrayType.getGenericComponentType());
+            return newInstance(componentClass, 0).getClass();
         }
         return Object.class;
     }

--- a/langchain4j-core/src/test/java/dev/langchain4j/internal/JsonSchemaElementUtilsTest.java
+++ b/langchain4j-core/src/test/java/dev/langchain4j/internal/JsonSchemaElementUtilsTest.java
@@ -187,9 +187,7 @@ class JsonSchemaElementUtilsTest {
         Map<String, Object> map = toMap(person, false);
 
         // then
-        assertThat(new ObjectMapper().writeValueAsString(map))
-                .isEqualToIgnoringWhitespace(
-                        """
+        assertThat(new ObjectMapper().writeValueAsString(map)).isEqualToIgnoringWhitespace("""
                 {
                    "type":"object",
                    "properties":{
@@ -221,9 +219,7 @@ class JsonSchemaElementUtilsTest {
         Map<String, Object> map = toMap(person, true);
 
         // then
-        assertThat(new ObjectMapper().writeValueAsString(map))
-                .isEqualToIgnoringWhitespace(
-                        """
+        assertThat(new ObjectMapper().writeValueAsString(map)).isEqualToIgnoringWhitespace("""
                 {
                    "type":"object",
                    "properties":{
@@ -245,8 +241,7 @@ class JsonSchemaElementUtilsTest {
     @Test
     void nativeSchemaToMap() throws JsonProcessingException {
         ObjectMapper mapper = new ObjectMapper();
-        String rawJsonSchema =
-                """
+        String rawJsonSchema = """
         {
             "additionalProperties": false,
             "type" : "object",
@@ -321,7 +316,8 @@ class JsonSchemaElementUtilsTest {
                                         .items(JsonObjectSchema.builder()
                                                 .addProperty(
                                                         "value",
-                                                        JsonObjectSchema.builder().build())
+                                                        JsonObjectSchema.builder()
+                                                                .build())
                                                 .required("value")
                                                 .build())
                                         .build())
@@ -356,7 +352,8 @@ class JsonSchemaElementUtilsTest {
                                         .items(JsonObjectSchema.builder()
                                                 .addProperty(
                                                         "value",
-                                                        JsonObjectSchema.builder().build())
+                                                        JsonObjectSchema.builder()
+                                                                .build())
                                                 .required("value")
                                                 .build())
                                         .build())
@@ -369,7 +366,8 @@ class JsonSchemaElementUtilsTest {
                                 "genericArrays",
                                 JsonArraySchema.builder()
                                         .items(JsonArraySchema.builder()
-                                                .items(JsonStringSchema.builder().build())
+                                                .items(JsonStringSchema.builder()
+                                                        .build())
                                                 .build())
                                         .build())
                         .required("wildcards", "typeVariables", "genericArrays")
@@ -397,7 +395,8 @@ class JsonSchemaElementUtilsTest {
                                 "nested",
                                 JsonArraySchema.builder()
                                         .items(JsonArraySchema.builder()
-                                                .items(JsonStringSchema.builder().build())
+                                                .items(JsonStringSchema.builder()
+                                                        .build())
                                                 .build())
                                         .build())
                         .required("nested")
@@ -409,16 +408,17 @@ class JsonSchemaElementUtilsTest {
 
         // given
         enum MyEnum {
-            A, B, C;
+            A,
+            B,
+            C;
         }
 
         // when
         JsonSchemaElement schema = jsonSchemaElementFrom(MyEnum.class);
 
         // then
-        assertThat(schema).isEqualTo(JsonEnumSchema.builder()
-                .enumValues("A", "B", "C")
-                .build());
+        assertThat(schema)
+                .isEqualTo(JsonEnumSchema.builder().enumValues("A", "B", "C").build());
     }
 
     @Test
@@ -426,7 +426,9 @@ class JsonSchemaElementUtilsTest {
 
         // given
         enum MyEnumWithToString {
-            A, B, C;
+            A,
+            B,
+            C;
 
             @Override
             public String toString() {
@@ -440,61 +442,48 @@ class JsonSchemaElementUtilsTest {
         JsonSchemaElement schema = jsonSchemaElementFrom(MyEnumWithToString.class);
 
         // then
-        assertThat(schema).isEqualTo(JsonEnumSchema.builder()
-                .enumValues("A", "B", "C")
-                .build());
+        assertThat(schema)
+                .isEqualTo(JsonEnumSchema.builder().enumValues("A", "B", "C").build());
     }
 
     @Test
     void shouldConvertJsonStringSchemaToMap() {
-        JsonStringSchema schema = JsonStringSchema.builder()
-                .description("string description")
-                .build();
+        JsonStringSchema schema =
+                JsonStringSchema.builder().description("string description").build();
 
         Map<String, Object> map = JsonSchemaElementUtils.toMap(schema);
 
-        assertThat(map)
-                .containsEntry("type", "string")
-                .containsEntry("description", "string description");
+        assertThat(map).containsEntry("type", "string").containsEntry("description", "string description");
     }
 
     @Test
     void shouldConvertJsonIntegerSchemaToMap() {
-        JsonIntegerSchema schema = JsonIntegerSchema.builder()
-                .description("integer description")
-                .build();
+        JsonIntegerSchema schema =
+                JsonIntegerSchema.builder().description("integer description").build();
 
         Map<String, Object> map = JsonSchemaElementUtils.toMap(schema);
 
-        assertThat(map)
-                .containsEntry("type", "integer")
-                .containsEntry("description", "integer description");
+        assertThat(map).containsEntry("type", "integer").containsEntry("description", "integer description");
     }
 
     @Test
     void shouldConvertJsonNumberSchemaToMap() {
-        JsonNumberSchema schema = JsonNumberSchema.builder()
-                .description("number description")
-                .build();
+        JsonNumberSchema schema =
+                JsonNumberSchema.builder().description("number description").build();
 
         Map<String, Object> map = JsonSchemaElementUtils.toMap(schema);
 
-        assertThat(map)
-                .containsEntry("type", "number")
-                .containsEntry("description", "number description");
+        assertThat(map).containsEntry("type", "number").containsEntry("description", "number description");
     }
 
     @Test
     void shouldConvertJsonBooleanSchemaToMap() {
-        JsonBooleanSchema schema = JsonBooleanSchema.builder()
-                .description("boolean description")
-                .build();
+        JsonBooleanSchema schema =
+                JsonBooleanSchema.builder().description("boolean description").build();
 
         Map<String, Object> map = JsonSchemaElementUtils.toMap(schema);
 
-        assertThat(map)
-                .containsEntry("type", "boolean")
-                .containsEntry("description", "boolean description");
+        assertThat(map).containsEntry("type", "boolean").containsEntry("description", "boolean description");
     }
 
     @Test
@@ -551,9 +540,8 @@ class JsonSchemaElementUtilsTest {
 
     @Test
     void shouldConvertJsonReferenceSchemaToMap() {
-        JsonReferenceSchema schema = JsonReferenceSchema.builder()
-                .reference("my-ref")
-                .build();
+        JsonReferenceSchema schema =
+                JsonReferenceSchema.builder().reference("my-ref").build();
 
         Map<String, Object> map = JsonSchemaElementUtils.toMap(schema);
 
@@ -565,8 +553,7 @@ class JsonSchemaElementUtilsTest {
         JsonAnyOfSchema schema = JsonAnyOfSchema.builder()
                 .anyOf(Arrays.asList(
                         JsonSchemaElementUtils.jsonSchemaElementFrom(String.class),
-                        JsonSchemaElementUtils.jsonSchemaElementFrom(Integer.class)
-                ))
+                        JsonSchemaElementUtils.jsonSchemaElementFrom(Integer.class)))
                 .description("anyOf description")
                 .build();
 
@@ -593,7 +580,8 @@ class JsonSchemaElementUtilsTest {
     @Test
     void shouldConvertJsonRawSchemaToMap() {
         Map<String, Object> rawMap = Map.of("foo", "bar");
-        JsonRawSchema schema = JsonRawSchema.builder().schema(Json.toJson(rawMap)).build();
+        JsonRawSchema schema =
+                JsonRawSchema.builder().schema(Json.toJson(rawMap)).build();
 
         Map<String, Object> map = JsonSchemaElementUtils.toMap(schema);
 

--- a/langchain4j-core/src/test/java/dev/langchain4j/internal/JsonSchemaElementUtilsTest.java
+++ b/langchain4j-core/src/test/java/dev/langchain4j/internal/JsonSchemaElementUtilsTest.java
@@ -292,6 +292,102 @@ class JsonSchemaElementUtilsTest {
         assertThat(isJsonArray(Iterable.class)).isTrue();
     }
 
+    static class GenericBox<T> {
+
+        T value;
+    }
+
+    static class HolderOfBoxedStrings {
+
+        List<GenericBox<String>> items;
+    }
+
+    @Test
+    void should_create_schema_when_collection_element_type_is_parameterized() {
+
+        // given
+        Class<HolderOfBoxedStrings> clazz = HolderOfBoxedStrings.class;
+
+        // when
+        JsonSchemaElement schema = jsonSchemaElementFrom(clazz, null, null, true, new LinkedHashMap<>());
+
+        // then — element class is the raw GenericBox; its T-typed `value` field erases to Object,
+        // which has no schema-relevant fields, so the box becomes an empty object.
+        assertThat(schema)
+                .isEqualTo(JsonObjectSchema.builder()
+                        .addProperty(
+                                "items",
+                                JsonArraySchema.builder()
+                                        .items(JsonObjectSchema.builder()
+                                                .addProperty(
+                                                        "value",
+                                                        JsonObjectSchema.builder().build())
+                                                .required("value")
+                                                .build())
+                                        .build())
+                        .required("items")
+                        .build());
+    }
+
+    static class HolderOfWildcardBoxes {
+
+        List<? extends GenericBox<String>> items;
+    }
+
+    @Test
+    void should_create_schema_when_collection_element_type_is_a_wildcard() {
+
+        // given
+        Class<HolderOfWildcardBoxes> clazz = HolderOfWildcardBoxes.class;
+
+        // when
+        JsonSchemaElement schema = jsonSchemaElementFrom(clazz, null, null, true, new LinkedHashMap<>());
+
+        // then — same shape as the parameterized case: the wildcard's upper bound resolves to GenericBox.
+        assertThat(schema)
+                .isEqualTo(JsonObjectSchema.builder()
+                        .addProperty(
+                                "items",
+                                JsonArraySchema.builder()
+                                        .items(JsonObjectSchema.builder()
+                                                .addProperty(
+                                                        "value",
+                                                        JsonObjectSchema.builder().build())
+                                                .required("value")
+                                                .build())
+                                        .build())
+                        .required("items")
+                        .build());
+    }
+
+    static class HolderOfNestedCollection {
+
+        List<List<String>> nested;
+    }
+
+    @Test
+    void should_create_nested_array_schema_for_collection_of_collections() {
+
+        // given
+        Class<HolderOfNestedCollection> clazz = HolderOfNestedCollection.class;
+
+        // when
+        JsonSchemaElement schema = jsonSchemaElementFrom(clazz, null, null, true, new LinkedHashMap<>());
+
+        // then — array-of-array-of-string, fully resolved through both Collection layers.
+        assertThat(schema)
+                .isEqualTo(JsonObjectSchema.builder()
+                        .addProperty(
+                                "nested",
+                                JsonArraySchema.builder()
+                                        .items(JsonArraySchema.builder()
+                                                .items(JsonStringSchema.builder().build())
+                                                .build())
+                                        .build())
+                        .required("nested")
+                        .build());
+    }
+
     @Test
     void should_create_schema_for_enum() {
 

--- a/langchain4j-core/src/test/java/dev/langchain4j/internal/JsonSchemaElementUtilsTest.java
+++ b/langchain4j-core/src/test/java/dev/langchain4j/internal/JsonSchemaElementUtilsTest.java
@@ -329,25 +329,29 @@ class JsonSchemaElementUtilsTest {
                         .build());
     }
 
-    static class HolderOfWildcardBoxes {
+    static class CombinedGenericsHolder<T extends CharSequence> {
 
-        List<? extends GenericBox<String>> items;
+        List<? extends GenericBox<String>> wildcards;
+
+        List<T> typeVariables;
+
+        List<T[]> genericArrays;
     }
 
     @Test
-    void should_create_schema_when_collection_element_type_is_a_wildcard() {
+    void should_resolve_wildcard_type_variable_and_generic_array_collection_elements() {
 
         // given
-        Class<HolderOfWildcardBoxes> clazz = HolderOfWildcardBoxes.class;
+        Class<CombinedGenericsHolder> clazz = CombinedGenericsHolder.class;
 
         // when
         JsonSchemaElement schema = jsonSchemaElementFrom(clazz, null, null, true, new LinkedHashMap<>());
 
-        // then — same shape as the parameterized case: the wildcard's upper bound resolves to GenericBox.
+        // then — every Collection element resolves through its respective rawClassOf branch.
         assertThat(schema)
                 .isEqualTo(JsonObjectSchema.builder()
                         .addProperty(
-                                "items",
+                                "wildcards",
                                 JsonArraySchema.builder()
                                         .items(JsonObjectSchema.builder()
                                                 .addProperty(
@@ -356,7 +360,19 @@ class JsonSchemaElementUtilsTest {
                                                 .required("value")
                                                 .build())
                                         .build())
-                        .required("items")
+                        .addProperty(
+                                "typeVariables",
+                                JsonArraySchema.builder()
+                                        .items(JsonStringSchema.builder().build())
+                                        .build())
+                        .addProperty(
+                                "genericArrays",
+                                JsonArraySchema.builder()
+                                        .items(JsonArraySchema.builder()
+                                                .items(JsonStringSchema.builder().build())
+                                                .build())
+                                        .build())
+                        .required("wildcards", "typeVariables", "genericArrays")
                         .build());
     }
 
@@ -374,7 +390,7 @@ class JsonSchemaElementUtilsTest {
         // when
         JsonSchemaElement schema = jsonSchemaElementFrom(clazz, null, null, true, new LinkedHashMap<>());
 
-        // then — array-of-array-of-string, fully resolved through both Collection layers.
+        // then — array-of-array-of-string
         assertThat(schema)
                 .isEqualTo(JsonObjectSchema.builder()
                         .addProperty(

